contracts/MagicatsHandler.sol

```diff
@@ -114,28 +113,36 @@ contract MagicatsHandler is AccessControlEnumerable, ERC721Enumerable {
      * @param magicatsIds - array of NFTs to withdraw
      * @dev rewards are forcefully claimed for all withdrawn nfts because they would be lost in limbo otherwise
      * this also burns the underlying nft.
+     *
+     * this function accounts for a possible case where this particular version of the magicatsHandler is retired from use,
+     * in this case, the NFTs are not in the strategy anymore but back in this contract for safe retrieval by their rightful owners
      */
-    function withdraw(uint256[] memory magicatsIds) external{
+    function withdraw(uint256[] memory magicatsIds) external {
         claimRewards(magicatsIds);
-        for(uint i; i < magicatsIds.length; i++){
+        for (uint256 i; i < magicatsIds.length; i = _uncheckedInc(i)) {
             require(_isApprovedOrOwner(msg.sender, magicatsIds[i]), "!approved");
             totalMp -= idToMagicat[magicatsIds[i]].manapoints;
 
-            if(IERC721(Magicats).ownerOf(magicatsIds[i]) == strategy){
+            if (IERC721(MAGICATS).ownerOf(magicatsIds[i]) == strategy) {
                 _burn(magicatsIds[i]);
```

Comment:
might be worth trying to refactor a bit so that burn is called the min. number of time and code is a bit easier to follow